apps/passport-client/src/localstorage.ts

```diff
@@ -12,11 +12,22 @@ import { SemaphoreSignaturePCD } from "@pcd/semaphore-signature-pcd";
 import { Identity } from "@semaphore-protocol/identity";
 import { z } from "zod";
 import { getPackages } from "./pcdPackages";
+import {
+  logAndUploadValidationErrors,
+  validatePCDCollection
+} from "./validateState";
+
+// validation points in this file
 
 const OLD_PCDS_KEY = "pcds"; // deprecated
 const COLLECTION_KEY = "pcd_collection";
 
 export async function savePCDs(pcds: PCDCollection): Promise<void> {
```

Comment:
I'd consider combining savePCDs+saveSubcriptions+saveSelf into a single function so the whole thing can be validated together, and to ensure they're all saved in sync with each other.  I actually think they should all be saved in a single localStorage key too.  This may be a bigger change than we want to make this week, though, so it might be fair to defer for risk reasons.

Reply:
I considered the extent to which it made sense to refactor vs. just add validation to what exists now, and I think we're on the same page re: what makes sense for this week.